Index: odhcpd-2022-03-22-860ca900/src/dhcpv4.c
===================================================================
--- odhcpd-2022-03-22-860ca900.orig/src/dhcpv4.c
+++ odhcpd-2022-03-22-860ca900/src/dhcpv4.c
@@ -1008,7 +1008,7 @@ void dhcpv4_handle_msg(void *addr, void
 		dhcpv4_put(&reply, &cookie, DHCPV4_OPT_DNSSERVER,
 				4 * iface->dhcpv4_dns_cnt, iface->dhcpv4_dns);
 
-	if (a->reqopts && iface->dhcpv4_ntp_cnt != 0) {
+	if (a && a->reqopts && iface->dhcpv4_ntp_cnt != 0) {
 		for(size_t opts = 0; a->reqopts[opts]; opts++) {
 			if (a->reqopts[opts] == DHCPV4_OPT_NTPSERVER) {
 				dhcpv4_put(&reply, &cookie, DHCPV4_OPT_NTPSERVER,
