Index: linux-5.10.125/drivers/tty/serial/Kconfig
===================================================================
--- linux-5.10.125.orig/drivers/tty/serial/Kconfig
+++ linux-5.10.125/drivers/tty/serial/Kconfig
@@ -1460,6 +1460,29 @@ config SERIAL_MEN_Z135
 	  This driver can also be build as a module. If so, the module will be called
 	  men_z135_uart.ko
 
+config SERIAL_CORTINA 
+        bool "CORTINA serial port support" 
+        select SERIAL_CORE 
+        help 
+          If you have a board based on a Cortina, you 
+          can enable its onboard serial port by enabling this option. 
+ 
+config SERIAL_CORTINA_CONSOLE 
+        bool "Console on CORTINA serial port" 
+        depends on SERIAL_CORTINA=y 
+        select SERIAL_CORE_CONSOLE 
+	select SERIAL_EARLYCON
+        help 
+          If you have enabled the serial port on the Cortina baseboard 
+          you can make it the console by answering Y to this option. 
+ 
+          Even if you say Y here, the currently visible virtual console 
+          (/dev/tty0) will still be used as the system console by default, but 
+          you can alter that using a kernel command line option such as 
+          "console=ttyS0". (Try "man bootparam" or see the documentation of
+          your boot loader (lilo or loadlin) about how to pass options to the
+          kernel at boot time.)
+
 config SERIAL_SPRD
 	tristate "Support for Spreadtrum serial"
 	select SERIAL_CORE
Index: linux-5.10.125/drivers/tty/serial/Makefile
===================================================================
--- linux-5.10.125.orig/drivers/tty/serial/Makefile
+++ linux-5.10.125/drivers/tty/serial/Makefile
@@ -90,6 +90,7 @@ obj-$(CONFIG_SERIAL_OWL)	+= owl-uart.o
 obj-$(CONFIG_SERIAL_RDA)	+= rda-uart.o
 obj-$(CONFIG_SERIAL_MILBEAUT_USIO) += milbeaut_usio.o
 obj-$(CONFIG_SERIAL_SIFIVE)	+= sifive.o
+obj-$(CONFIG_SERIAL_CORTINA)    += serial_cortina.o
 
 # GPIOLIB helpers for modem control lines
 obj-$(CONFIG_SERIAL_MCTRL_GPIO)	+= serial_mctrl_gpio.o
Index: linux-5.10.125/include/uapi/linux/serial_core.h
===================================================================
--- linux-5.10.125.orig/include/uapi/linux/serial_core.h
+++ linux-5.10.125/include/uapi/linux/serial_core.h
@@ -279,4 +279,7 @@
 /* Freescale LINFlexD UART */
 #define PORT_LINFLEXUART	122
 
+/* Cortina SoC - place here to prevent merge conflict later */
+#define PORT_CORTINA    199
+
 #endif /* _UAPILINUX_SERIAL_CORE_H */
