--- a/drivers/char/Kconfig
+++ b/drivers/char/Kconfig
@@ -130,6 +130,40 @@ config POWERNV_OP_PANEL
 
 	  If unsure, say M here to build it as a module called powernv-op-panel.
 
+config RTK_PE
+	bool "Packet Engine"
+	default n
+
+config RTK_PE_STATIC_BUFFER_RESERVE
+	bool "Reserve a static buffer for PE"
+	depends on RTK_PE
+	default y
+	help
+	Say Y here to reserve a static buffer for PE.
+
+config CORTINA_HWSEM
+	tristate "Cortina SoC Hardware semaphore module"
+	depends on ARCH_CORTINA
+	default n
+	help
+	  Say Y here to enable support for hardware semephore module provided by the
+	  Cortina SoC.
+
+config REALTEK_IPC2RCPU
+	tristate "Realtek SoC IPC-to-RCPU module"
+	depends on ARCH_CORTINA && RTK_PE
+	default n
+	help
+	  Say Y here to enable support for IPC to RCPU module.
+	  
+config CORTINA_PELOG
+	tristate "Cortina SoC pelog module"
+	depends on ARCH_CORTINA && RTK_PE
+	default y
+	help
+	  Say Y here to enable support for pelog module provided by the
+	  Cortina SoC.
+
 source "drivers/char/ipmi/Kconfig"
 
 config DS1620
--- a/drivers/char/Makefile
+++ b/drivers/char/Makefile
@@ -3,6 +3,12 @@
 # Makefile for the kernel character device drivers.
 #
 
+EXTRA_CFLAGS += -I$(srctree)/include/soc/cortina/
+obj-$(CONFIG_RTK_PE) += ca_pe_ctrl.o
+obj-$(CONFIG_REALTEK_IPC2RCPU)	+= ca_soft_ipc.o ca_pe_ctrl.o
+obj-$(CONFIG_CORTINA_PELOG)	+= ca_pelog.o ca_pe_ctrl.o
+obj-$(CONFIG_CORTINA_HWSEM)	+= ca_hwsem.o
+
 obj-y				+= mem.o random.o
 obj-$(CONFIG_TTY_PRINTK)	+= ttyprintk.o
 obj-y				+= misc.o
