--- a/include/linux/skbuff.h
+++ b/include/linux/skbuff.h
@@ -41,6 +41,15 @@
 #include <linux/netfilter/nf_conntrack_common.h>
 #endif
 
+#if defined(CONFIG_RTK_L34_FLEETCONNTRACK_ENABLE)
+#include "../../drivers/net/ethernet/realtek/rtl86900/FleetConntrackDriver/rtk_fc_skb.h"
+#endif
+#if defined(CONFIG_RTL_ETH_RECYCLED_SKB)
+#include "../../drivers/net/ethernet/realtek/rtl86900/nicPool/re_recycleconf.h"	
+#endif
+#if defined(CONFIG_WNIC_RECYCLED_SKB)
+#include "../../drivers/net/wireless/realtek/wnic_skb_pool/recycle_conf.h"
+#endif
 /* The interface for checksum offload between the stack and networking drivers
  * is as follows...
  *
@@ -225,7 +234,13 @@
 /* Maximum value in skb->csum_level */
 #define SKB_MAX_CSUM_LEVEL	3
 
+#ifdef CONFIG_ARCH_CORTINA
+#include <soc/cortina/ca_offload_info.h>
+/* QM requests 128-byte alignment */
+#define SKB_DATA_ALIGN(X)	ALIGN(X, 1<<7)
+#else
 #define SKB_DATA_ALIGN(X)	ALIGN(X, SMP_CACHE_BYTES)
+#endif
 #define SKB_WITH_OVERHEAD(X)	\
 	((X) - SKB_DATA_ALIGN(sizeof(struct skb_shared_info)))
 #define SKB_MAX_ORDER(X, ORDER) \
@@ -858,7 +873,6 @@ struct sk_buff {
 #ifdef CONFIG_TLS_DEVICE
 	__u8			decrypted:1;
 #endif
-	__u8			gro_skip:1;
 
 #ifdef CONFIG_NET_SCHED
 	__u16			tc_index;	/* traffic control index */
@@ -921,8 +935,91 @@ struct sk_buff {
 	/* only useable after checking ->active_extensions != 0 */
 	struct skb_ext		*extensions;
 #endif
+	/* ===Realtek private fields start===*/
+	__u32 			__private_header_start[0];
+
+//#if defined( CONFIG_RTL_HARDWARE_MULTICAST) || defined(CONFIG_RTL865X_LANPORT_RESTRICTION) ||defined(CONFIG_RTL_QOS_VLANID_SUPPORT)||defined (CONFIG_RTL_VLAN_8021Q)
+	__u16			srcPort;
+	__u16			srcVlanId:12;
+//#endif
+
+#if defined(CONFIG_RTL_ETH_RECYCLED_SKB) || defined(CONFIG_WNIC_RECYCLED_SKB)
+#if defined(CONFIG_RTL_ETH_RECYCLED_SKB)
+	__u8			recyclable:RTL_POOL_NUM_SHIFT_BITS;
+	__u8			frag_list_recycle:1;	//1: the shar_info->frag_list point to skb and so on
+#endif
+#if defined(CONFIG_WNIC_RECYCLED_SKB)
+	__u8			wnic_recyclable:WNIC_POOL_NUM_SHIFT_BITS;
+#endif
+#if	defined(CONFIG_RTL_ETH_RECYCLED_SKB_DEBUG)
+	__u8            double_check_free:1;
+	struct sk_buff  *skb_bak;
+	unsigned char	*head_bak;
+#endif
+	__u8			prepared_recycle:1;		//1: ready to free, 0:not yet
+#if defined(RTL_LL_LINK_LIST_RECYCLE_SKB) || defined(WNIC_LL_LINK_LIST_RECYCLE_SKB)
+	struct llist_node recy_llnode;
+#endif
+#endif
+
+#if defined(CONFIG_RTK_SKB_MARK2)
+	union {
+		__u64		mark2;
+		__u64		reserved_tailroom2;
+	};
+#endif
+
+	struct net_device		*from_dev;				/* Device we arrived on */
+
+#if defined(CONFIG_RTL8192CD) || defined(CONFIG_RTL8192CD_MODULE) || defined(CONFIG_RTL8852AE_BACKPORTS)
+	/*This field is only needed by RtL8190 Driver.FIX ME!!!*/
+        unsigned char   __unused;
+#endif
+
+#if defined(CONFIG_RTK_L34_ENABLE)
+	__u8				fwd_drop;				/* 1: dropped by fwdEngine, 0:normal case */
+#endif
+
+#if defined(CONFIG_CTC_AWIFI_KERNEL_JITUAN)
+	__u32			usr_define;	/* Bit 0 : awifi purpose */
+#endif
+
+#if defined(CONFIG_RTL_SMUX_DEV)
+#ifndef SMUX_MAX_TAGS
+#define SMUX_MAX_TAGS 4
+#endif
+	__u8			smux_filter_pkt;
+	__s8			reserv_vlan_count;
+	__s8			original_vlan_count; //0: initial, -1: untagged, 1~n: valid n VLAN headers
+	__u32			reserv_vlan_header[SMUX_MAX_TAGS];
+	__u32			original_vlan_header[SMUX_MAX_TAGS];
+#endif
+
+#if defined(CONFIG_RTK_L34_FLEETCONNTRACK_ENABLE)
+	rtk_fc_ingress_data_t fcIngressData;
+	__u32 stat_ip;
+#endif
+
+	/* private data used in Cortina NI driver */
+#ifdef CONFIG_ARCH_CORTINA
+	__u8			is_from_l3qm;
+	__u8			fbm_pool_id;
+#ifdef CONFIG_CORTINA_KERNEL_HOOK
+#define IFNAMSIZ	16
+	__u16 			vid;
+	__u8 			br_member_dev_name[IFNAMSIZ+1];
+#endif
+#endif
+
+	__u32 			__private_header_end[0];
+	/* ===Realtek private fields End===*/
 };
 
+
+#ifdef CONFIG_RTK_FC_TCP_SPI_SUPPORT
+bool rtk_skip_nf_conntrack_in(struct sk_buff *skb);
+#endif
+
 #ifdef __KERNEL__
 /*
  *	Handling routines are only of interest to the kernel
@@ -2677,7 +2774,11 @@ static inline int pskb_network_may_pull(
  * NET_IP_ALIGN(2) + ethernet_header(14) + IP_header(20/40) + ports(8)
  */
 #ifndef NET_SKB_PAD
-#define NET_SKB_PAD	max(64, L1_CACHE_BYTES)
+#if defined(CONFIG_ARCH_CORTINA) || defined(CONFIG_RTK_SOC_RTL8198D)
+#define NET_SKB_PAD     max(32, 1<<7)
+#else
+#define NET_SKB_PAD	max(32, L1_CACHE_BYTES)
+#endif
 #endif
 
 int ___pskb_trim(struct sk_buff *skb, unsigned int len);
@@ -2710,10 +2811,6 @@ static inline int pskb_trim(struct sk_bu
 	return (len < skb->len) ? __pskb_trim(skb, len) : 0;
 }
 
-extern struct sk_buff *__netdev_alloc_skb_ip_align(struct net_device *dev,
-		unsigned int length, gfp_t gfp);
-
-
 /**
  *	pskb_trim_unique - remove end from a paged unique (not cloned) buffer
  *	@skb: buffer to alter
@@ -2845,6 +2942,16 @@ static inline struct sk_buff *dev_alloc_
 }
 
 
+static inline struct sk_buff *__netdev_alloc_skb_ip_align(struct net_device *dev,
+		unsigned int length, gfp_t gfp)
+{
+	struct sk_buff *skb = __netdev_alloc_skb(dev, length + NET_IP_ALIGN, gfp);
+
+	if (NET_IP_ALIGN && skb)
+		skb_reserve(skb, NET_IP_ALIGN);
+	return skb;
+}
+
 static inline struct sk_buff *netdev_alloc_skb_ip_align(struct net_device *dev,
 		unsigned int length)
 {
--- a/net/core/skbuff.c
+++ b/net/core/skbuff.c
@@ -80,6 +80,21 @@
 
 #include "datagram.h"
 
+#ifdef CONFIG_RTL865X_ETH_PRIV_SKB
+extern int is_rtl865x_eth_priv_buf(unsigned char *head);
+extern void free_rtl865x_eth_priv_buf(unsigned char *head);
+extern int is_rtl865x_critical_eth_priv_buf(unsigned char *head);
+extern void free_rtl865x_critical_eth_priv_buf(unsigned char *head);
+#endif
+#if defined(CONFIG_RTL_ETH_RECYCLED_SKB)
+int recycle_rtl_eth_buf(struct sk_buff *skb);
+void recycle_skb_clean(struct sk_buff *skb);
+#endif
+#if defined(CONFIG_WNIC_RECYCLED_SKB)
+int recycle_wnic_buf(struct sk_buff *skb);
+void wnic_recycle_skb_clean(struct sk_buff *skb);
+#endif
+
 struct kmem_cache *skbuff_head_cache __ro_after_init;
 static struct kmem_cache *skbuff_fclone_cache __ro_after_init;
 #ifdef CONFIG_SKB_EXTENSIONS
@@ -224,6 +239,10 @@ struct sk_buff *__alloc_skb(unsigned int
 	 * the tail pointer in struct sk_buff!
 	 */
 	memset(skb, 0, offsetof(struct sk_buff, tail));
+#if 1 //clear RTK private header
+	memset(&skb->__private_header_start, 0, offsetof(struct sk_buff, __private_header_end) -
+											offsetof(struct sk_buff, __private_header_start));
+#endif
 	/* Account for allocated memory : skb + skb->head */
 	skb->truesize = SKB_TRUESIZE(size);
 	skb->pfmemalloc = pfmemalloc;
@@ -262,6 +281,135 @@ nodata:
 }
 EXPORT_SYMBOL(__alloc_skb);
 
+#if defined(CONFIG_RTL865X_ETH_PRIV_SKB) || defined(CONFIG_RTL_ETH_RECYCLED_SKB)
+#ifndef RTL_PRIV_DATA_SIZE
+#define RTL_PRIV_DATA_SIZE	128
+#endif
+#endif
+
+#if defined(CONFIG_RTL865X_ETH_PRIV_SKB)
+struct sk_buff *dev_alloc_8190_skb(unsigned char *data, int size)
+{
+    struct sk_buff *skb;
+	struct skb_shared_info *shinfo;
+
+ 	struct kmem_cache *cache;
+ 	cache = skbuff_head_cache;
+	/* Get the HEAD */
+    skb = kmem_cache_alloc(cache, GFP_ATOMIC & ~__GFP_DMA);
+    if (!skb)
+            goto out;
+
+    memset(skb, 0, offsetof(struct sk_buff, tail));
+#if 1 //clear RTK private header
+	memset(&skb->__private_header_start, 0, offsetof(struct sk_buff, __private_header_end) -
+										offsetof(struct sk_buff, __private_header_start));
+#endif
+    refcount_set(&skb->users, 1);
+    skb->head = data;
+    skb->data = data;
+    skb->tail = data;
+
+	size = SKB_DATA_ALIGN(size+RTL_PRIV_DATA_SIZE+NET_SKB_PAD);
+
+    skb->end  = data + size;
+    skb->end = ((unsigned long)skb->end + SMP_CACHE_BYTES - 1) & ~(SMP_CACHE_BYTES - 1);
+    skb->truesize = size + sizeof(struct sk_buff);
+
+    /* make sure we initialize shinfo sequentially */
+    shinfo = skb_shinfo(skb);
+	memset(shinfo, 0, offsetof(struct skb_shared_info, dataref));
+    atomic_set(&shinfo->dataref, 1);
+
+#ifdef CONFIG_RTK_VOIP_VLAN_ID
+    skb->rx_vlan = 0;
+    skb->rx_wlan = 0;
+    skb->priority = 0;
+#endif
+
+#if defined(CONFIG_RTL_ISP_MULTI_WAN_SUPPORT)
+	skb->from_dev = NULL;
+	skb->vlan_member = 0;
+#endif
+
+#if defined(CONFIG_RTL_IP_POLICY_ROUTING_SUPPORT)
+	skb->switch_port = NULL;
+#endif
+
+#if defined(CONFIG_RTL_CUSTOM_PASSTHRU) && defined (CONFIG_RTL_VLAN_8021Q) && defined(CONFIG_RTL_8021Q_VLAN_SUPPORT_SRC_TAG)
+	skb->passThruFlag = 0;
+#endif
+
+//#if defined(CONFIG_RTL_HARDWARE_MULTICAST) ||defined(CONFIG_RTL_QOS_VLANID_SUPPORT)||defined (CONFIG_RTL_VLAN_8021Q)
+	skb->srcPort=0xFFFF;
+	skb->srcVlanId=0;
+//#endif
+
+#if	defined(CONFIG_RTL_QOS_8021P_SUPPORT)
+	skb->srcVlanPriority=0;
+#endif
+
+#if defined(CONFIG_RTL_DSCP_IPTABLE_CHECK)			
+	skb->original_dscp=0;
+ #endif
+
+ #if defined(CONFIG_RTL_VLANPRI_IPTABLE_CHECK)
+ 	skb->original_vlanpri=0;
+ #endif
+
+#if defined(CONFIG_NETFILTER_XT_MATCH_PHYPORT) || defined(CONFIG_RTL_FAST_FILTER) || defined(CONFIG_RTL_QOS_PATCH) || defined(CONFIG_RTK_VOIP_QOS)|| defined(CONFIG_RTK_VLAN_WAN_TAG_SUPPORT) ||defined(CONFIG_RTL_MAC_FILTER_CARE_INPORT) ||defined(CONFIG_RTL_HW_QOS_SUPPORT_WLAN) || defined(CONFIG_RTL_DNS_TRAP) || defined(CONFIG_RTL_NIC_QUEUE)|| defined(CONFIG_RTL_CUSTOM_PASSTHRU)
+	skb->srcPhyPort=0xFF;
+	skb->dstPhyPort=0xFF;
+#endif
+
+#if defined(CONFIG_RTK_VLAN_SUPPORT)
+	skb->tag.v = 0;
+#if defined(CONFIG_RTK_BRIDGE_VLAN_SUPPORT)
+	skb->src_info = NULL;
+#endif
+#endif
+
+#if defined(CONFIG_RTL_VLAN_PASSTHROUGH_SUPPORT)
+	memset(skb->vlan_passthrough_saved_tag, 0, 4);
+#endif
+
+#if defined(CONFIG_RTL_VLAN_8021Q)
+	memset(skb->linux_vlan_src_tag, 0, 4);
+#endif
+
+#if defined(CONFIG_RTL_DNS_TRAP) || defined(CONFIG_RTL_HTTP_REDIRECT)
+	skb->dns_trap=0;
+	skb->http_redirect=0;
+	skb->is_dns_pkt = 0;
+#endif
+
+#if defined (CONFIG_RTL_LOCAL_PUBLIC)
+	skb->srcLocalPublicIp=0;
+	skb->fromLocalPublic=0;
+	skb->toLocalPublic=0;
+#endif
+
+#if defined(CONFIG_RTL_HW_QOS_SUPPORT) && defined(CONFIG_RTL_SW_QUEUE_DECISION_PRIORITY)
+	skb->decision_bitmap = 0;
+	memset(skb->mark_ext, 0, sizeof(skb->mark_ext));
+#endif
+#if defined(CONFIG_RTL_IPTABLES_FAST_PATH)
+	skb->inDev=NULL;
+#endif
+
+#ifdef CONFIG_RTK_VOIP_VLAN_ID
+	skb_reserve(skb, RTL_PRIV_DATA_SIZE+4); // for VLAN TAG insertion
+#else
+	skb_reserve(skb, RTL_PRIV_DATA_SIZE);
+#endif
+
+        return skb;
+
+out:
+        return NULL;
+}
+#endif // CONFIG_RTL8190_PRIV_SKB
+
 /* Caller must provide SKB that is memset cleared */
 static struct sk_buff *__build_skb_around(struct sk_buff *skb,
 					  void *data, unsigned int frag_size)
@@ -319,6 +467,10 @@ struct sk_buff *__build_skb(void *data,
 		return NULL;
 
 	memset(skb, 0, offsetof(struct sk_buff, tail));
+#if 1 //clear RTK private header
+	memset(&skb->__private_header_start, 0, offsetof(struct sk_buff, __private_header_end) -
+											offsetof(struct sk_buff, __private_header_start));
+#endif
 
 	return __build_skb_around(skb, data, frag_size);
 }
@@ -611,10 +763,34 @@ static void skb_clone_fraglist(struct sk
 		skb_get(list);
 }
 
+#ifdef CONFIG_RTL8190_PRIV_SKB
+extern int is_rtl8190_priv_buf(unsigned char *head);
+extern void free_rtl8190_priv_buf(unsigned char *head);
+#endif
+
 static void skb_free_head(struct sk_buff *skb)
 {
 	unsigned char *head = skb->head;
 
+#if defined(CONFIG_RTL_ETH_RECYCLED_SKB)		
+	if(skb->recyclable)		
+		BUG();		
+	else
+#endif
+#if defined(CONFIG_WNIC_RECYCLED_SKB)
+	if(skb->wnic_recyclable)
+		BUG();
+	else
+#endif
+#if defined(CONFIG_RTL865X_ETH_PRIV_SKB)
+	if (is_rtl865x_eth_priv_buf(skb->head)) {
+		free_rtl865x_eth_priv_buf(skb->head);
+	}
+	else if(is_rtl865x_critical_eth_priv_buf(skb->head)) {
+		free_rtl865x_critical_eth_priv_buf(skb->head);
+	}
+	else
+#endif
 	if (skb->head_frag)
 		skb_free_frag(head);
 	else
@@ -631,6 +807,21 @@ static void skb_release_data(struct sk_b
 			      &shinfo->dataref))
 		return;
 
+#if defined(CONFIG_WNIC_RECYCLED_SKB)
+	//recycle skb should be free later
+	if(skb->wnic_recyclable){
+		skb->prepared_recycle = 1;
+		return;
+	}
+#endif
+#if defined(CONFIG_RTL_ETH_RECYCLED_SKB)	
+	//recycle skb should be free later	
+	if(skb->recyclable){	
+		skb->prepared_recycle = 1;	
+		return;	
+	}	
+#endif
+
 	for (i = 0; i < shinfo->nr_frags; i++)
 		__skb_frag_unref(&shinfo->frags[i]);
 
@@ -707,7 +898,33 @@ static void skb_release_all(struct sk_bu
 
 void __kfree_skb(struct sk_buff *skb)
 {
+#if defined(CONFIG_RTL_ETH_RECYCLED_SKB) || defined(CONFIG_WNIC_RECYCLED_SKB)
+	if(skb->cloned==0){
+		//return to queue and exit
+#if defined(CONFIG_RTL_ETH_RECYCLED_SKB)
+		if(recycle_rtl_eth_buf(skb)) 
+			return;
+#endif
+#if	defined(CONFIG_WNIC_RECYCLED_SKB)
+		if(recycle_wnic_buf(skb))
+			return;
+#endif
+	}
+#endif
+
 	skb_release_all(skb);
+#if defined(CONFIG_RTL_ETH_RECYCLED_SKB) || defined(CONFIG_WNIC_RECYCLED_SKB)
+	if(skb->prepared_recycle){
+#if	defined(CONFIG_WNIC_RECYCLED_SKB)
+		if(recycle_wnic_buf(skb))
+			return;
+#endif
+#if	defined(CONFIG_RTL_ETH_RECYCLED_SKB)
+		if(recycle_rtl_eth_buf(skb))
+			return;
+#endif
+	}else
+#endif
 	kfree_skbmem(skb);
 }
 EXPORT_SYMBOL(__kfree_skb);
@@ -861,7 +1078,19 @@ void consume_skb(struct sk_buff *skb)
 {
 	if (!skb_unref(skb))
 		return;
-
+#if defined(CONFIG_RTL_ETH_RECYCLED_SKB) || defined(CONFIG_WNIC_RECYCLED_SKB)
+	if(skb->cloned==0){
+		//return to queue and exit
+#if defined(CONFIG_RTL_ETH_RECYCLED_SKB)
+		if(recycle_rtl_eth_buf(skb)) 
+			return;
+#endif
+#if	defined(CONFIG_WNIC_RECYCLED_SKB)
+		if(recycle_wnic_buf(skb))
+			return;
+#endif
+	}
+#endif
 	trace_consume_skb(skb);
 	__kfree_skb(skb);
 }
@@ -970,6 +1199,11 @@ static void __copy_skb_header(struct sk_
 	memcpy(&new->headers_start, &old->headers_start,
 	       offsetof(struct sk_buff, headers_end) -
 	       offsetof(struct sk_buff, headers_start));
+#if 1 //copy RTK private header
+	memcpy(&new->__private_header_start, &old->__private_header_start,
+		   offsetof(struct sk_buff, __private_header_end) -
+		   offsetof(struct sk_buff, __private_header_start));
+#endif
 	CHECK_SKB_FIELD(protocol);
 	CHECK_SKB_FIELD(csum);
 	CHECK_SKB_FIELD(hash);
@@ -1502,6 +1736,15 @@ void skb_copy_header(struct sk_buff *new
 {
 	__copy_skb_header(new, old);
 
+#if defined(CONFIG_RTL_ETH_RECYCLED_SKB)	
+	//for copied skb recyclable bit should always be zero.	
+	new->recyclable=0;	
+#endif
+#if defined(CONFIG_WNIC_RECYCLED_SKB)
+	//for copied skb wnic_recyclable bit should always be zero.
+	new->wnic_recyclable=0;
+#endif
+
 	skb_shinfo(new)->gso_size = skb_shinfo(old)->gso_size;
 	skb_shinfo(new)->gso_segs = skb_shinfo(old)->gso_segs;
 	skb_shinfo(new)->gso_type = skb_shinfo(old)->gso_type;
@@ -5307,7 +5550,13 @@ void skb_scrub_packet(struct sk_buff *sk
 
 	ipvs_reset(skb);
 	skb->mark = 0;
+#ifdef CONFIG_RTK_SKB_MARK2
+	skb->mark2 = 0;
+#endif
 	skb->tstamp = 0;
+#ifdef CONFIG_RTK_SKB_MARK2
+	skb->mark2 = 0;
+#endif
 }
 EXPORT_SYMBOL_GPL(skb_scrub_packet);
 
@@ -6398,3 +6647,15 @@ free_now:
 }
 EXPORT_SYMBOL(__skb_ext_put);
 #endif /* CONFIG_SKB_EXTENSIONS */
+
+#ifdef CONFIG_RTK_FC_TCP_SPI_SUPPORT
+bool rtk_skip_nf_conntrack_in(struct sk_buff *skb)
+{
+	if(skb && skb->fcIngressData.skip_nf_conntrack_in)
+		return true;
+
+	return false;
+}
+EXPORT_SYMBOL(rtk_skip_nf_conntrack_in);
+#endif
+
--- a/include/net/netfilter/nf_conntrack.h
+++ b/include/net/netfilter/nf_conntrack.h
@@ -94,6 +94,9 @@ struct nf_conn {
 
 #if defined(CONFIG_NF_CONNTRACK_MARK)
 	u_int32_t mark;
+#if defined(CONFIG_RTK_SKB_MARK2)
+        u_int64_t mark2;
+#endif
 #endif
 
 #ifdef CONFIG_NF_CONNTRACK_SECMARK
--- a/net/Kconfig
+++ b/net/Kconfig
@@ -114,6 +114,12 @@ config NETWORK_SECMARK
 	  to nfmark, but designated for security purposes.
 	  If you are unsure how to answer this question, answer N.
 
+config RTK_SKB_MARK2
+       bool "Realtek skb mark2"
+       default y if GPON_FEATURE || EPON_FEATURE
+       help
+         This enables skb mark2 marking of network packets.
+
 config NET_PTP_CLASSIFY
 	def_bool n
 
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -3171,7 +3171,21 @@ void __dev_kfree_skb_irq(struct sk_buff
 
 	if (unlikely(!skb))
 		return;
-
+#if defined(CONFIG_RTL_ETH_RECYCLED_SKB) || defined(CONFIG_WNIC_RECYCLED_SKB)
+	if(skb->cloned==0){
+		//return to queue and exit
+#if defined(CONFIG_RTL_ETH_RECYCLED_SKB)
+		int recycle_rtl_eth_buf(struct sk_buff *skb);
+		if(recycle_rtl_eth_buf(skb)) 
+			return;
+#endif
+#if defined(CONFIG_WNIC_RECYCLED_SKB)
+		int recycle_wnic_buf(struct sk_buff *skb);
+		if(recycle_wnic_buf(skb))
+			return;
+#endif
+	}
+#endif
 	if (likely(refcount_read(&skb->users) == 1)) {
 		smp_rmb();
 		refcount_set(&skb->users, 0);
@@ -3189,6 +3203,21 @@ EXPORT_SYMBOL(__dev_kfree_skb_irq);
 
 void __dev_kfree_skb_any(struct sk_buff *skb, enum skb_free_reason reason)
 {
+#if defined(CONFIG_RTL_ETH_RECYCLED_SKB) || defined(CONFIG_WNIC_RECYCLED_SKB)
+	if(skb && skb->cloned==0){
+		//return to queue and exit
+#if defined(CONFIG_RTL_ETH_RECYCLED_SKB)
+		int recycle_rtl_eth_buf(struct sk_buff *skb);
+		if(recycle_rtl_eth_buf(skb)) 
+			return;
+#endif
+#if defined(CONFIG_WNIC_RECYCLED_SKB)
+		int recycle_wnic_buf(struct sk_buff *skb);
+		if(recycle_wnic_buf(skb))
+			return;
+#endif
+	}
+#endif
 	if (in_irq() || irqs_disabled())
 		__dev_kfree_skb_irq(skb, reason);
 	else
@@ -4941,6 +4970,25 @@ int netif_rx(struct sk_buff *skb)
 	int ret;
 
 	trace_netif_rx_entry(skb);
+	
+#ifdef CONFIG_RTL_ETH_RECYCLED_SKB
+	struct sk_buff *recycle_skb_swap(struct sk_buff *skb);
+	if(skb->recyclable)
+	{
+		skb=recycle_skb_swap(skb);
+		if(skb==NULL)
+			return NET_RX_SUCCESS;
+	}
+#endif
+#if	defined(CONFIG_WNIC_RECYCLED_SKB)
+	if(skb->wnic_recyclable)
+	{
+		struct sk_buff *wnic_recycle_skb_swap(struct sk_buff *skb);
+		skb=wnic_recycle_skb_swap(skb);
+		if(skb==NULL)
+			return NET_RX_SUCCESS;
+	}
+#endif
 
 	ret = netif_rx_internal(skb);
 	trace_netif_rx_exit(ret);
@@ -6246,6 +6294,25 @@ gro_result_t napi_gro_receive(struct nap
 
 	skb_mark_napi_id(skb, napi);
 	trace_napi_gro_receive_entry(skb);
+	
+#ifdef CONFIG_RTL_ETH_RECYCLED_SKB
+	if(skb->recyclable)
+	{
+		struct sk_buff *recycle_skb_swap(struct sk_buff *skb);
+		skb=recycle_skb_swap(skb);
+		if(skb==NULL)
+			return GRO_DROP;
+	}
+#endif
+#if defined(CONFIG_WNIC_RECYCLED_SKB)
+	if(skb->wnic_recyclable)
+	{
+		struct sk_buff *wnic_recycle_skb_swap(struct sk_buff *skb);
+		skb=wnic_recycle_skb_swap(skb);
+		if(skb==NULL)
+			return GRO_DROP;
+	}
+#endif
 
 	skb_gro_reset_offset(skb, 0);
 
