Index: linux-5.10.138/net/core/skbuff.c
===================================================================
--- linux-5.10.138.orig/net/core/skbuff.c
+++ linux-5.10.138/net/core/skbuff.c
@@ -706,22 +706,6 @@ skb_fail:
 }
 EXPORT_SYMBOL(__napi_alloc_skb);
 
-struct sk_buff *__netdev_alloc_skb_ip_align(struct net_device *dev,
-		unsigned int length, gfp_t gfp)
-{
-	struct sk_buff *skb = __netdev_alloc_skb(dev, length + NET_IP_ALIGN, gfp);
-
-#ifdef CONFIG_ETHERNET_PACKET_MANGLE
-	if (dev && (dev->priv_flags & IFF_NO_IP_ALIGN))
-		return skb;
-#endif
-
-	if (NET_IP_ALIGN && skb)
-		skb_reserve(skb, NET_IP_ALIGN);
-	return skb;
-}
-EXPORT_SYMBOL(__netdev_alloc_skb_ip_align);
-
 void skb_add_rx_frag(struct sk_buff *skb, int i, struct page *page, int off,
 		     int size, unsigned int truesize)
 {
