From ccb4541fe3113dfc0f7c1abc2b878f7c9b06d3ce Mon Sep 17 00:00:00 2001
From: Qubot <1445788683@qq.com>
Date: Thu, 29 Jan 2026 16:19:15 +0800
Subject: [PATCH 12/12] update lan led io

---
 .../drivers/soc/realtek/board_rtl8198d.c         | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/rtk_openwrt/target/linux/realtek_bb/files-5.10/drivers/soc/realtek/board_rtl8198d.c b/rtk_openwrt/target/linux/realtek_bb/files-5.10/drivers/soc/realtek/board_rtl8198d.c
index 385b86f0..d7d24871 100755
--- a/rtk_openwrt/target/linux/realtek_bb/files-5.10/drivers/soc/realtek/board_rtl8198d.c
+++ b/rtk_openwrt/target/linux/realtek_bb/files-5.10/drivers/soc/realtek/board_rtl8198d.c
@@ -69,8 +69,8 @@ static void board_led_init(void){
 	led_cfg_10_100m.ledEnable[LED_CONFIG_SPD10] = ENABLED;
   	
 	rtk_led_operation_set(LED_OP_PARALLEL);
-	rtk_led_parallelEnable_set(2, ENABLED);  // port 0
-	rtk_led_parallelEnable_set(3, ENABLED);  // port 1
+	rtk_led_parallelEnable_set(14, ENABLED);  // port 0
+	rtk_led_parallelEnable_set(13, ENABLED);  // port 1
 	rtk_led_parallelEnable_set(17, ENABLED); // port 2
 	rtk_led_parallelEnable_set(5, ENABLED);  // port 3
 	rtk_led_parallelEnable_set(4, ENABLED);  // port 0
@@ -78,8 +78,8 @@ static void board_led_init(void){
 	rtk_led_parallelEnable_set(9, ENABLED);  // port 2
 	rtk_led_parallelEnable_set(8, ENABLED);  // port 3
   	
-	rtk_led_config_set(2,  LED_TYPE_UTP0,  &led_cfg_giga);
-	rtk_led_config_set(3,  LED_TYPE_UTP1,  &led_cfg_giga);
+	rtk_led_config_set(14,  LED_TYPE_UTP0,  &led_cfg_giga);
+	rtk_led_config_set(13,  LED_TYPE_UTP1,  &led_cfg_giga);
 	rtk_led_config_set(17, LED_TYPE_UTP2,  &led_cfg_giga);
 	rtk_led_config_set(5,  LED_TYPE_UTP3,  &led_cfg_giga);
 
@@ -103,13 +103,13 @@ static void board_led_init(void){
 	led_cfg.ledEnable[LED_CONFIG_SPD10] = ENABLED;
   	
 	rtk_led_operation_set(LED_OP_PARALLEL);
-	rtk_led_parallelEnable_set(2, ENABLED);  // port 0
-	rtk_led_parallelEnable_set(3, ENABLED);  // port 1
+	rtk_led_parallelEnable_set(14, ENABLED);  // port 0
+	rtk_led_parallelEnable_set(13, ENABLED);  // port 1
 	rtk_led_parallelEnable_set(17, ENABLED); // port 2
 	rtk_led_parallelEnable_set(5, ENABLED);  // port 3
   	
-	rtk_led_config_set(2,  LED_TYPE_UTP0,  &led_cfg);
-	rtk_led_config_set(3,  LED_TYPE_UTP1,  &led_cfg);
+	rtk_led_config_set(14,  LED_TYPE_UTP0,  &led_cfg);
+	rtk_led_config_set(13,  LED_TYPE_UTP1,  &led_cfg);
 	rtk_led_config_set(17, LED_TYPE_UTP2,  &led_cfg);
 	rtk_led_config_set(5,  LED_TYPE_UTP3,  &led_cfg);
 #endif
-- 
2.34.1

