diff --git a/rtk_openwrt/vendor/Realtek_ap/conf2203/V0J4G_98D_8832D_8192XB_8221B_MULTI_WAN_EASYMESH/dts/rtl8198d_engboard.dts b/rtk_openwrt/vendor/Realtek_ap/conf2203/V0J4G_98D_8832D_8192XB_8221B_MULTI_WAN_EASYMESH/dts/rtl8198d_engboard.dts
index c39ff75..2979d8d 100755
--- a/rtk_openwrt/vendor/Realtek_ap/conf2203/V0J4G_98D_8832D_8192XB_8221B_MULTI_WAN_EASYMESH/dts/rtl8198d_engboard.dts
+++ b/rtk_openwrt/vendor/Realtek_ap/conf2203/V0J4G_98D_8832D_8192XB_8221B_MULTI_WAN_EASYMESH/dts/rtl8198d_engboard.dts
@@ -58,6 +58,40 @@
 		pci1_gpio_rst = <&gpio1 7 GPIO_ACTIVE_HIGH>;
 	};
 	
+	i2c_gpio_0: i2c-gpio-0 {
+		compatible = "i2c-gpio";
+		sda-gpios = <&gpio0 1 GPIO_ACTIVE_HIGH>;
+		scl-gpios = <&gpio0 0 GPIO_ACTIVE_HIGH>;
+		i2c-gpio,delay-us = <2>;
+		#address-cells = <1>;
+		#size-cells = <0>;
+	};
+
+	i2c_gpio_1: i2c-gpio-1 {
+   		compatible = "i2c-gpio";
+   		sda-gpios = <&gpio0 8 GPIO_ACTIVE_HIGH>;
+   		scl-gpios = <&gpio0 9 GPIO_ACTIVE_HIGH>;
+   		i2c-gpio,delay-us = <5>;
+    	#address-cells = <1>;
+    	#size-cells = <0>;
+
+  		status = "okay";
+    	eeprom@57 {
+        	compatible = "atmel,24c02";
+        	reg = <0x57>;
+        	pagesize = <8>;
+    	};
+	};
+	
+};
+
+&gpio1 {
+    eeprom_wp_disable {
+    	gpio-hog;
+    	gpios = <30 GPIO_ACTIVE_HIGH>;
+    	output-low;
+    	line-name = "eeprom-wp";
+	};
 };
 
 &gic {
