diff --git a/rtk_openwrt/package/realtek/rtk_base/src/fc_api.c b/rtk_openwrt/package/realtek/rtk_base/src/fc_api.c
index 49f7aa1..decd2e2 100755
--- a/rtk_openwrt/package/realtek/rtk_base/src/fc_api.c
+++ b/rtk_openwrt/package/realtek/rtk_base/src/fc_api.c
@@ -197,11 +197,16 @@ int rtk_fc_ctrl_init(void)
 	if (deviceType == DEVICE_TYPE_SFU || deviceType == DEVICE_TYPE_HYBRID) //hybrid and SFU mode
 		va_cmd(BIN_SH, 2, 1, "-c", "echo 1 > " FC_PROC"/ctrl/l2_port_moving_check_with_vlan");
 
-	wanPhyPort = rtk_port_get_wan_phyID();
-	if (wanPhyPort != -1) {
-		sprintf(buf, "echo 0x%x > " FC_PROC"/ctrl/wan_port_mask", (1<<wanPhyPort));		// change fc wan port mask
-		va_cmd(BIN_SH, 2, 1, "-c", buf);		// change fc wan port mask
-	}
+//	wanPhyPort = rtk_port_get_wan_phyID();
+//	if (wanPhyPort != -1) {
+//		sprintf(buf, "echo 0x%x > " FC_PROC"/ctrl/wan_port_mask", (1<<wanPhyPort));		// change fc wan port mask
+//		va_cmd(BIN_SH, 2, 1, "-c", buf);		// change fc wan port mask
+//	}
+
+	/* Remove Port 0 from LAN mask */
+    va_cmd(BIN_SH, 2, 1, "-c", "echo 0x3e > " FC_PROC"/ctrl/mac_port_mask_without_cpu");
+    /* Set Port 0 as the only WAN port */
+    va_cmd(BIN_SH, 2, 1, "-c", "echo 0x01 > " FC_PROC"/ctrl/wan_port_mask");
 
 #ifdef CONFIG_RTK_L34_FC_IPI_NIC_TX
 	rtk_fc_nic_smp_dispatch(RTK_FC_SMP_DISPATCH_TX);
