--- .config	2026-02-24 17:50:45.173602638 +0800
+++ .config-1	2026-02-10 16:24:34.955039363 +0800
@@ -2043,7 +2043,7 @@
 CONFIG_BUSYBOX_CONFIG_IPCRM=y
 CONFIG_BUSYBOX_CONFIG_IPCS=y
 # CONFIG_BUSYBOX_CONFIG_LOSETUP is not set
-# CONFIG_BUSYBOX_CONFIG_LSPCI is not set
+CONFIG_BUSYBOX_CONFIG_LSPCI=y
 # CONFIG_BUSYBOX_CONFIG_LSUSB is not set
 # CONFIG_BUSYBOX_CONFIG_MDEV is not set
 # CONFIG_BUSYBOX_CONFIG_MESG is not set
@@ -2222,7 +2222,7 @@
 # CONFIG_BUSYBOX_CONFIG_NBDCLIENT is not set
 # CONFIG_BUSYBOX_CONFIG_NC is not set
 # CONFIG_BUSYBOX_CONFIG_NETCAT is not set
-# CONFIG_BUSYBOX_CONFIG_NETMSG is not set
+CONFIG_BUSYBOX_CONFIG_NETMSG=y
 CONFIG_BUSYBOX_CONFIG_NETSTAT=y
 CONFIG_BUSYBOX_CONFIG_FEATURE_NETSTAT_WIDE=y
 CONFIG_BUSYBOX_CONFIG_FEATURE_NETSTAT_PRG=y
@@ -2313,7 +2313,7 @@
 CONFIG_BUSYBOX_CONFIG_KILLALL=y
 # CONFIG_BUSYBOX_CONFIG_KILLALL5 is not set
 # CONFIG_BUSYBOX_CONFIG_LSOF is not set
-# CONFIG_BUSYBOX_CONFIG_MPSTAT is not set
+CONFIG_BUSYBOX_CONFIG_MPSTAT=y
 # CONFIG_BUSYBOX_CONFIG_NMETER is not set
 CONFIG_BUSYBOX_CONFIG_PGREP=y
 # CONFIG_BUSYBOX_CONFIG_PKILL is not set
@@ -2452,7 +2452,7 @@
 # CONFIG_PACKAGE_libgomp is not set
 CONFIG_PACKAGE_libpthread=y
 CONFIG_PACKAGE_librt=y
-CONFIG_PACKAGE_libstdcpp=y
+# CONFIG_PACKAGE_libstdcpp is not set
 # CONFIG_PACKAGE_libthread-db is not set
 # CONFIG_PACKAGE_libubsan is not set
 CONFIG_PACKAGE_logd=y
@@ -2983,12 +2983,12 @@
 #
 # I2C support
 #
-# CONFIG_PACKAGE_kmod-i2c-algo-bit is not set
+CONFIG_PACKAGE_kmod-i2c-algo-bit=y
 # CONFIG_PACKAGE_kmod-i2c-algo-pca is not set
 # CONFIG_PACKAGE_kmod-i2c-algo-pcf is not set
-# CONFIG_PACKAGE_kmod-i2c-core is not set
+CONFIG_PACKAGE_kmod-i2c-core=y
 # CONFIG_PACKAGE_kmod-i2c-designware-pci is not set
-# CONFIG_PACKAGE_kmod-i2c-gpio is not set
+CONFIG_PACKAGE_kmod-i2c-gpio=y
 # CONFIG_PACKAGE_kmod-i2c-mux is not set
 # CONFIG_PACKAGE_kmod-i2c-mux-gpio is not set
 # CONFIG_PACKAGE_kmod-i2c-mux-pca9541 is not set
@@ -3393,7 +3393,7 @@
 # CONFIG_PACKAGE_kmod-ebt_mark2 is not set
 # CONFIG_PACKAGE_kmod-echo is not set
 # CONFIG_PACKAGE_kmod-eeprom-93cx6 is not set
-# CONFIG_PACKAGE_kmod-eeprom-at24 is not set
+CONFIG_PACKAGE_kmod-eeprom-at24=y
 # CONFIG_PACKAGE_kmod-eeprom-at25 is not set
 # CONFIG_PACKAGE_kmod-extGphy is not set
 # CONFIG_PACKAGE_kmod-gpio-beeper is not set
@@ -3428,6 +3428,8 @@
 # CONFIG_PACKAGE_kmod-ramoops is not set
 # CONFIG_PACKAGE_kmod-random-core is not set
 # CONFIG_PACKAGE_kmod-reed-solomon is not set
+CONFIG_PACKAGE_kmod-regmap-core=y
+CONFIG_PACKAGE_kmod-regmap-i2c=y
 # CONFIG_PACKAGE_kmod-rtc-ds1307 is not set
 # CONFIG_PACKAGE_kmod-rtc-ds1374 is not set
 # CONFIG_PACKAGE_kmod-rtc-ds1672 is not set
@@ -3440,7 +3442,7 @@
 # CONFIG_PACKAGE_kmod-rtc-rs5c372a is not set
 # CONFIG_PACKAGE_kmod-rtc-rx8025 is not set
 # CONFIG_PACKAGE_kmod-rtc-s35390a is not set
-# CONFIG_PACKAGE_kmod-rtk_hostspeedup is not set
+CONFIG_PACKAGE_kmod-rtk_hostspeedup=y
 CONFIG_PACKAGE_kmod-rtk_igmp_hook=y
 # CONFIG_PACKAGE_kmod-rtl8221b is not set
 # CONFIG_PACKAGE_kmod-sdhci is not set
@@ -4846,7 +4848,7 @@
 # CONFIG_PACKAGE_libavahi-dbus-support is not set
 # CONFIG_PACKAGE_libavahi-nodbus-support is not set
 # CONFIG_PACKAGE_libbfd is not set
-# CONFIG_PACKAGE_libblkid is not set
+CONFIG_PACKAGE_libblkid=y
 CONFIG_PACKAGE_libblobmsg-json=y
 CONFIG_PACKAGE_libbpf=y
 # CONFIG_PACKAGE_libbsd is not set
@@ -4880,7 +4882,7 @@
 # CONFIG_PACKAGE_libesmtp is not set
 # CONFIG_PACKAGE_libestr is not set
 # CONFIG_PACKAGE_libev is not set
-# CONFIG_PACKAGE_libevdev is not set
+CONFIG_PACKAGE_libevdev=y
 # CONFIG_PACKAGE_libevent2 is not set
 # CONFIG_PACKAGE_libevent2-core is not set
 # CONFIG_PACKAGE_libevent2-extra is not set
@@ -4977,7 +4979,7 @@
 # CONFIG_PACKAGE_libmodbus is not set
 # CONFIG_PACKAGE_libmosquitto-nossl is not set
 # CONFIG_PACKAGE_libmosquitto-ssl is not set
-# CONFIG_PACKAGE_libmount is not set
+CONFIG_PACKAGE_libmount=y
 # CONFIG_PACKAGE_libmpdclient is not set
 # CONFIG_PACKAGE_libmpeg2 is not set
 # CONFIG_PACKAGE_libmpg123 is not set
@@ -5093,7 +5095,7 @@
 # CONFIG_PACKAGE_libslang2-mod-varray is not set
 # CONFIG_PACKAGE_libslang2-mod-zlib is not set
 # CONFIG_PACKAGE_libslang2-modules is not set
-# CONFIG_PACKAGE_libsmartcols is not set
+CONFIG_PACKAGE_libsmartcols=y
 # CONFIG_PACKAGE_libsndfile is not set
 # CONFIG_PACKAGE_libsoc is not set
 # CONFIG_PACKAGE_libsocks is not set
@@ -5126,7 +5128,7 @@
 # CONFIG_PACKAGE_libuci-lua is not set
 # CONFIG_PACKAGE_libuci2 is not set
 CONFIG_PACKAGE_libuclient=y
-# CONFIG_PACKAGE_libudev-zero is not set
+CONFIG_PACKAGE_libudev-zero=y
 # CONFIG_PACKAGE_libudns is not set
 # CONFIG_PACKAGE_libuecc is not set
 # CONFIG_PACKAGE_libugpio is not set
@@ -5136,12 +5138,12 @@
 # CONFIG_PACKAGE_libupnpp is not set
 # CONFIG_PACKAGE_liburcu is not set
 # CONFIG_PACKAGE_liburing is not set
-# CONFIG_PACKAGE_libusb-1.0 is not set
+CONFIG_PACKAGE_libusb-1.0=y
 # CONFIG_PACKAGE_libusb-compat is not set
 # CONFIG_PACKAGE_libustream-mbedtls is not set
 CONFIG_PACKAGE_libustream-openssl=y
 # CONFIG_PACKAGE_libustream-wolfssl is not set
-# CONFIG_PACKAGE_libuuid is not set
+CONFIG_PACKAGE_libuuid=y
 CONFIG_PACKAGE_libuv=y
 # CONFIG_PACKAGE_libuwifi is not set
 # CONFIG_PACKAGE_libv4l is not set
@@ -5218,7 +5220,6 @@
 #
 CONFIG_PACKAGE_luci=y
 # CONFIG_PACKAGE_luci-nginx is not set
-# CONFIG_PACKAGE_luci-rtk-generic is not set
 # CONFIG_PACKAGE_luci-ssl is not set
 # CONFIG_PACKAGE_luci-ssl-nginx is not set
 CONFIG_PACKAGE_luci-ssl-openssl=y
@@ -5277,20 +5278,6 @@
 # CONFIG_PACKAGE_luci-mod-dashboard is not set
 CONFIG_PACKAGE_luci-mod-network=y
 CONFIG_PACKAGE_luci-mod-rpc=y
-# CONFIG_PACKAGE_luci-mod-rtk-admin is not set
-# CONFIG_PACKAGE_luci-mod-rtk-advance is not set
-# CONFIG_PACKAGE_luci-mod-rtk-bridging is not set
-# CONFIG_PACKAGE_luci-mod-rtk-diag is not set
-# CONFIG_PACKAGE_luci-mod-rtk-generic is not set
-# CONFIG_PACKAGE_luci-mod-rtk-lan is not set
-# CONFIG_PACKAGE_luci-mod-rtk-multiap is not set
-# CONFIG_PACKAGE_luci-mod-rtk-route is not set
-# CONFIG_PACKAGE_luci-mod-rtk-service is not set
-# CONFIG_PACKAGE_luci-mod-rtk-statistic is not set
-# CONFIG_PACKAGE_luci-mod-rtk-status is not set
-# CONFIG_PACKAGE_luci-mod-rtk-voip is not set
-# CONFIG_PACKAGE_luci-mod-rtk-wan is not set
-CONFIG_PACKAGE_luci-mod-rtk-wlan=y
 CONFIG_PACKAGE_luci-mod-status=y
 CONFIG_PACKAGE_luci-mod-system=y
 # end of 2. Modules
@@ -5348,7 +5335,6 @@
 # CONFIG_PACKAGE_luci-app-olsr is not set
 # CONFIG_PACKAGE_luci-app-olsr-services is not set
 # CONFIG_PACKAGE_luci-app-olsr-viz is not set
-# CONFIG_PACKAGE_luci-app-omcproxy is not set
 # CONFIG_PACKAGE_luci-app-openvpn is not set
 # CONFIG_PACKAGE_luci-app-openwisp is not set
 CONFIG_PACKAGE_luci-app-opkg=y
@@ -5360,25 +5346,6 @@
 # CONFIG_PACKAGE_luci-app-radicale is not set
 # CONFIG_PACKAGE_luci-app-radicale2 is not set
 # CONFIG_PACKAGE_luci-app-rp-pppoe-server is not set
-# CONFIG_PACKAGE_luci-app-rtk-br-grouping is not set
-# CONFIG_PACKAGE_luci-app-rtk-dhcpv4 is not set
-# CONFIG_PACKAGE_luci-app-rtk-dhcpv6 is not set
-# CONFIG_PACKAGE_luci-app-rtk-firewall is not set
-# CONFIG_PACKAGE_luci-app-rtk-gpon is not set
-CONFIG_PACKAGE_luci-app-rtk-igmpproxy=y
-# CONFIG_PACKAGE_luci-app-rtk-igmpsnooping is not set
-# CONFIG_PACKAGE_luci-app-rtk-ipsec is not set
-# CONFIG_PACKAGE_luci-app-rtk-l2tp is not set
-CONFIG_PACKAGE_luci-app-rtk-mcproxy=y
-CONFIG_PACKAGE_luci-app-rtk-multi-ap-release-ap=y
-# CONFIG_PACKAGE_luci-app-rtk-omcproxy is not set
-# CONFIG_PACKAGE_luci-app-rtk-pptp is not set
-# CONFIG_PACKAGE_luci-app-rtk-qos is not set
-# CONFIG_PACKAGE_luci-app-rtk-radvd is not set
-# CONFIG_PACKAGE_luci-app-rtk-samba is not set
-# CONFIG_PACKAGE_luci-app-rtk-tr069 is not set
-# CONFIG_PACKAGE_luci-app-rtk-tr369 is not set
-# CONFIG_PACKAGE_luci-app-rtk-upnp is not set
 # CONFIG_PACKAGE_luci-app-samba4 is not set
 # CONFIG_PACKAGE_luci-app-ser2net is not set
 # CONFIG_PACKAGE_luci-app-shadowsocks-libev is not set
@@ -5419,7 +5386,6 @@
 # CONFIG_PACKAGE_luci-theme-material is not set
 # CONFIG_PACKAGE_luci-theme-openwrt is not set
 CONFIG_PACKAGE_luci-theme-openwrt-2020=y
-# CONFIG_PACKAGE_luci-theme-rtk-generic is not set
 # end of 4. Themes
 
 #
@@ -5427,7 +5393,6 @@
 #
 # CONFIG_PACKAGE_luci-proto-3g is not set
 # CONFIG_PACKAGE_luci-proto-bonding is not set
-# CONFIG_PACKAGE_luci-proto-gre is not set
 # CONFIG_PACKAGE_luci-proto-hnet is not set
 # CONFIG_PACKAGE_luci-proto-ipip is not set
 CONFIG_PACKAGE_luci-proto-ipv6=y
@@ -5567,10 +5532,6 @@
 # CONFIG_PACKAGE_luci-i18n-opkg-vi is not set
 # CONFIG_PACKAGE_luci-i18n-opkg-zh-cn is not set
 # CONFIG_PACKAGE_luci-i18n-opkg-zh-tw is not set
-# CONFIG_PACKAGE_luci-i18n-rtk-multi-ap-release-ap-de is not set
-# CONFIG_PACKAGE_luci-i18n-rtk-multi-ap-release-ap-ms is not set
-# CONFIG_PACKAGE_luci-i18n-rtk-multi-ap-release-ap-zh-cn is not set
-# CONFIG_PACKAGE_luci-i18n-rtk-multi-ap-release-ap-zh-tw is not set
 # end of LuCI
 
 #
@@ -6050,7 +6011,7 @@
 CONFIG_PACKAGE_ip-full=y
 # CONFIG_PACKAGE_ip-tiny is not set
 # CONFIG_PACKAGE_lldpd is not set
-CONFIG_PACKAGE_mcproxy=y
+# CONFIG_PACKAGE_mcproxy is not set
 # CONFIG_PACKAGE_mrmctl is not set
 # CONFIG_PACKAGE_mwan3 is not set
 # CONFIG_PACKAGE_nstat is not set
@@ -6422,7 +6383,7 @@
 # CONFIG_PACKAGE_esniper is not set
 # CONFIG_PACKAGE_etherwake is not set
 # CONFIG_PACKAGE_etherwake-nfqueue is not set
-# CONFIG_PACKAGE_ethtool is not set
+CONFIG_PACKAGE_ethtool=y
 # CONFIG_PACKAGE_ethtool-full is not set
 # CONFIG_PACKAGE_fail2ban is not set
 # CONFIG_PACKAGE_fakeidentd is not set
@@ -6453,7 +6414,7 @@
 # CONFIG_PACKAGE_iftop is not set
 # CONFIG_PACKAGE_iiod is not set
 # CONFIG_PACKAGE_iperf is not set
-# CONFIG_PACKAGE_iperf3 is not set
+CONFIG_PACKAGE_iperf3=y
 # CONFIG_PACKAGE_iperf3-ssl is not set
 # CONFIG_PACKAGE_ipip is not set
 # CONFIG_PACKAGE_ipset is not set
@@ -6782,6 +6743,7 @@
 # Disc
 #
 # CONFIG_PACKAGE_autopart is not set
+# CONFIG_PACKAGE_blkdiscard is not set
 # CONFIG_PACKAGE_blkid is not set
 # CONFIG_PACKAGE_blockdev is not set
 # CONFIG_PACKAGE_cfdisk is not set
@@ -6795,7 +6757,7 @@
 # CONFIG_PACKAGE_gdisk is not set
 # CONFIG_PACKAGE_hd-idle is not set
 # CONFIG_PACKAGE_hdparm is not set
-# CONFIG_PACKAGE_lsblk is not set
+CONFIG_PACKAGE_lsblk=y
 # CONFIG_PACKAGE_lvm2 is not set
 # CONFIG_PACKAGE_lvm2-selinux is not set
 # CONFIG_PACKAGE_mdadm is not set
@@ -7493,7 +7455,7 @@
 # CONFIG_PACKAGE_unshare is not set
 # CONFIG_PACKAGE_usb-modeswitch is not set
 # CONFIG_PACKAGE_usbids is not set
-# CONFIG_PACKAGE_usbutils is not set
+CONFIG_PACKAGE_usbutils=y
 # CONFIG_PACKAGE_uuidd is not set
 # CONFIG_PACKAGE_uuidgen is not set
 # CONFIG_PACKAGE_uvcdynctrl is not set
