--- a/os_dep/linux/os_intfs.c
+++ b/os_dep/linux/os_intfs.c
@@ -90,6 +90,12 @@ int rtw_net_set_mac_address(struct net_d
 	_rtw_memcpy(adapter_mac_addr(padapter), sa->sa_data, ETH_ALEN); /* set mac addr to adapter */
 	_rtw_memcpy(pnetdev->dev_addr, sa->sa_data, ETH_ALEN); /* set mac addr to net_device */
 
+#ifdef CONFIG_IOCTL_CFG80211
+	if(is_primary_adapter(padapter)){
+		_rtw_memcpy(padapter->rtw_wdev->wiphy->perm_addr, sa->sa_data, ETH_ALEN); /* set mac addr to wiphy perm_addr */
+	}
+#endif
+
 	rtw_ps_deny(padapter, PS_DENY_IOCTL);
 	LeaveAllPowerSaveModeDirect(padapter); /* leave PS mode for guaranteeing to access hw register successfully */
 
