Index: netifd-2022-08-25-76d2d41b/system-linux.c
===================================================================
--- netifd-2022-08-25-76d2d41b.orig/system-linux.c
+++ netifd-2022-08-25-76d2d41b/system-linux.c
@@ -862,10 +862,16 @@ system_bridge_set_wireless(struct device
 	bool mcast_to_ucast = dev->wireless_ap;
 	bool hairpin;
 
+#if 1 // RTK: dont enable hairpin_mode and multicast_to_unicast when config AP mode
+	mcast_to_ucast = false;
+	if (bridge->settings.flags & DEV_OPT_MULTICAST_TO_UNICAST &&
+	    bridge->settings.multicast_to_unicast)
+		mcast_to_ucast = true;
+#else
 	if (bridge->settings.flags & DEV_OPT_MULTICAST_TO_UNICAST &&
 	    !bridge->settings.multicast_to_unicast)
 		mcast_to_ucast = false;
-
+#endif
 	hairpin = mcast_to_ucast || dev->wireless_proxyarp;
 	if (dev->wireless_isolate)
 		hairpin = false;
