--- a/iwlib.c
+++ b/iwlib.c
@@ -1138,7 +1138,7 @@
 void
 iw_print_bitrate(char *	buffer,
 		 int	buflen,
-		 int	bitrate)
+		 unsigned int	bitrate)
 {
   double	rate = bitrate;
   char		scale;
--- a/iwlib.h
+++ b/iwlib.h
@@ -346,7 +346,7 @@
 void
 	iw_print_bitrate(char *	buffer,
 			 int	buflen,
-			 int	bitrate);
+			 unsigned int	bitrate);
 /* ---------------------- POWER SUBROUTINES ----------------------- */
 int
 	iw_dbm2mwatt(int	in);
